### PR TITLE
Hosting Panel: Add selector and banner for Business plan upgrade

### LINKED ISSUE #36488: Hosting: Add a banner if folks don't have a business plan [3]
In Phase 3, the Hosting Section is visible to Simple and Atomic sites. (Exclude VIP and Jetpack standalone).

Please put this behind a feature flag until we're ready to turn this on

Clicking on upgrade should take us to checkout with a Business plan in cart. We should also fire a related tracks event to capture this.

Once folks complete checkout, we should redirect to the section, where behavior should match phase 2.

Screen Shot 2019-12-03 at 8 55 21 AM

### PR DESCRIPTION
**Pull request by jeraldjuice**
### Summary

This PR adds an upgrade banner to the top of the Hosting Configuration section if the site is not on a Business plan, in preparation to expose visibility of the section to Simple sites.

This work will be encapsulated behind a feature flag when #38155 is merged and the visibility selector for the Hosting section performs a feature flag check for `hosting/all-sites`.

<hr>

### Screenshots

**Upgrade banner displayed on Hosting section**

Screen Shot 2019-12-04 at 11 15 49 PM

**Additional button added to Thank You page after upgrade**
Screen Shot 2019-12-09 at 4 41 24 PM

<hr>

### Testing

* Start up Calypso with the `hosting/all-sites` flag running:
`ENABLE_FEATURES=hosting/all-sites npm start`
* **Banner display:**
    * Go to a Simple Site. Ensure that the 'Upgrade' banner is visible.
    * Click on the banner and ensure that you are directed to the shopping cart with the business plan added
    * Go to a site on a Business plan and ensure that the banner is not visible.
* **Thank you page button:**
    * Go to a Simple Site. Click on the Upgrade banner to upgrade your site.
    * After checking out, ensure that a second button is displayed that says "Return to Hosting" and that this button returns you to the Hosting Configuration panel for the site.
    * Test the non-hosting upgrade flow by copying the Thank You page URL, navigating to the Plugins section (or Themes section) of your site, then navigating to the Thank You page URL from your previous checkout. Ensure that the "Back to Hosting" button does not appear.

Fixes #36488

### CODE REVIEW

**Comment by gwwar** [comment]
> use update/hosting-display-all rebased on top of this branch:

💖 Nice testing instructions!

File: client/state/selectors/is-site-on-business-plan.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * @returns { boolean } If the site is on a Business plan
+ */
+const isSiteOnBusinessPlan = ( state, siteId ) => {
+	const currentPlan = getCurrentPlan( state, siteId );
+
+	if ( ! currentPlan ) {
+		return false;
+	}
+
+	return isBusinessPlan( currentPlan.productSlug );
```
Comment (gwwar) [inline client/state/selectors/is-site-on-business-plan.js]: We need to check for a Business plan or eCommerce plan

Comment (gwwar) [reply client/state/selectors/is-site-on-business-plan.js]: The existing https://github.com/Automattic/wp-calypso/blob/57261aa3f0b3398e720a6ae91a6c03b88028e62c/client/state/selectors/is-site-on-atomic-plan.js should do the trick. So no need for a new selector.

File: client/my-sites/hosting/main.js [hunk 2]
```diff
@@ -74,6 +78,15 @@ class Hosting extends Component {
 		const sftpPhpMyAdminFeaturesEnabled =
 			isEnabled( 'hosting/sftp-phpmyadmin' ) && siteId > 155000000;
 
+		const getUpgradeBanner = () => (
+			<Banner
+				title={ translate( 'Upgrade to the Business plan to access all hosting features' ) }
+				event="calypso_hosting_configuration_upgrade_click"
+				href={ `/checkout/${ siteId }/business` }
+				plan={ PLAN_BUSINESS }
```
Comment (gwwar) [inline client/my-sites/hosting/main.js]: A nudge to business makes sense (as ecommerce starts as atomic) 👍

File: client/state/selectors/is-site-on-business-plan.js (hunk 1 above)
Comment (jeraldjuice) [reply client/state/selectors/is-site-on-business-plan.js]: Ahhh thanks! :)

**Comment by gwwar** [comment]
Pretty close @jeraldjuice ! Is there any chance we can update the redirect after purchase back to the hosting section? While testing I was directed to a thank you page. There might be an existing mechanism if we 🔍 a bit.

**Comment by jeraldjuice** [comment]
@gwwar sure! i think that there is. i shall do that

**Comment by kwight** [comment]
Rebased, now that #38155 and related PRs are in 👍

**Comment by kwight** [comment]
> Is there any chance we can update the redirect after purchase back to the hosting section? While testing I was directed to a thank you page.

I dunno – the thank you page for the business upgrade is super useful and appreciative 🤔 What if we kept the redirect as-is, but added another button to allow going back to hosting?..

Screen Shot 2019-12-06 at 1 05 55 PM

cc/ @olaolusoga

**Comment by gwwar** [comment]
> the thank you page for the business upgrade is super useful and appreciative

Redirecting back to the hosting section was a request from @apeatling, but happy to defer if he changes his mind.

**Comment by apeatling** [comment]
I think this page is fine if there is a button.

**Comment by gwwar** [comment]
Wanted to leave a note that Andy would be happy with either a redirect or adding another section in the thank you page so folks can make it back to the /hosting-config section, so the choice is up to you on what you prefer @jeraldjuice

**Comment by jeraldjuice** [comment]
Button added on Thank You page after upgrade from Hosting section. Updated testing instructions and screenshots in PR description 👍

File: client/state/ui/checkout/test/reducer.js [hunk 3]
```diff
@@ -110,6 +110,14 @@ describe( 'reducer', () => {
 			expect( state ).toBe( 'themes' );
 		} );
 
+		test( 'should return hosting for hosting', () => {
+			const state = upgradeIntent( 'blastendedskrewt', {
```
Comment (gwwar) [inline client/state/ui/checkout/test/reducer.js]: For folks reading the first argument is what the current upgrade intent is, usually a section name. 

Also had to look this one up, lol:
Screen Shot 2019-12-09 at 1 55 46 PM

File: client/my-sites/checkout/checkout-thank-you/header.jsx [hunk 4]
```diff
@@ -252,6 +253,16 @@ export class CheckoutThankYouHeader extends PureComponent {
...
 	};
 
+	visitSiteHostingSettings = event => {
+		event.preventDefault();
+
+		const { selectedSite } = this.props;
+
+		this.props.recordTracksEvent( 'calypso_thank_you_back_to_hosting' );
+
+		window.location.href = `/hosting-config/${ selectedSite.slug }`;
```
Comment (gwwar) [inline client/my-sites/checkout/checkout-thank-you/header.jsx]: non-blocking and I know this is copy pasta from the page, but there is a potential race condition here where the tracks event may not complete before the browser navigates to a new section.

In a follow up janitorial, we can update these actions to use the `NAVIGATE` redux action, that is [decorated with meta analytics](https://github.com/Automattic/wp-calypso/blob/c1d1fac0fad54945ddb8b2a119e498fb07e97ed6/client/state/analytics/actions/with-analytics.js#L17), so two side-effects occur:

 - The analytics middleware sees the meta value on the action and fires the related tracks event.

https://github.com/Automattic/wp-calypso/blob/c1d1fac0fad54945ddb8b2a119e498fb07e97ed6/client/state/analytics/middleware.js#L75

- The navigation middleware, triggers a `page` router call

https://github.com/Automattic/wp-calypso/blob/c1d1fac0fad54945ddb8b2a119e498fb07e97ed6/client/state/navigation/middleware.js#L43

**Pull request merged**